File: 02-linked-lists/02_merge_two_sorted_lists/merge_two_sorted_lists.c
/**
 * Problem: Merge Two Sorted Lists
 * LeetCode: #21 (Easy)
 * 
 * Approach: Iterative Dummy Head Splicing
 * Time Complexity: O(N + M)
 * Space Complexity: O(1)
 */

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* createNode(int val) {
    struct ListNode *newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode dummy;
    struct ListNode *tail = &dummy;
    dummy.next = NULL;

    while (list1 != NULL && list2 != NULL) {
        if (list1->val <= list2->val) {
            tail->next = list1;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }

    if (list1 != NULL) tail->next = list1;
    else tail->next = list2;

    return dummy.next;
}

void printList(struct ListNode* head) {
    while (head) {
        printf("%d -> ", head->val);
        head = head->next;
    }
    printf("NULL\n");
}

int main(void) {
    printf("[C] Merge Two Sorted Lists Test\n");
    // List 1: 1 -> 2 -> 4
    struct ListNode *l1 = createNode(1);
    l1->next = createNode(2);
    l1->next->next = createNode(4);

    // List 2: 1 -> 3 -> 4
    struct ListNode *l2 = createNode(1);
    l2->next = createNode(3);
    l2->next->next = createNode(4);

    struct ListNode *merged = mergeTwoLists(l1, l2);
    printf("Merged List: ");
    printList(merged);
    return 0;
}

File: 02-linked-lists/02_merge_two_sorted_lists/merge_two_sorted_lists.py
"""
Problem: Merge Two Sorted Lists
LeetCode: #21 (Easy)

Approach: Iterative Dummy Node
Time Complexity: O(N + M)
Space Complexity: O(1)
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def merge_two_lists(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode(0)
    tail = dummy

    while l1 and l2:
        if l1.val <= l2.val:
            tail.next = l1
            l1 = l1.next
        else:
            tail.next = l2
            l2 = l2.next
        tail = tail.next

    tail.next = l1 if l1 else l2
    return dummy.next

def to_list(node):
    res = []
    while node:
        res.append(node.val)
        node = node.next
    return res

if __name__ == "__main__":
    print("[Python] Merge Two Sorted Lists Test")
    l1 = ListNode(1, ListNode(2, ListNode(4)))
    l2 = ListNode(1, ListNode(3, ListNode(4)))
    merged = merge_two_lists(l1, l2)
    result = to_list(merged)
    print("Merged Result:", result)
    assert result == [1, 1, 2, 3, 4, 4]
    print("Test passed!")
